### ISSUE TITLE
0.10.0 catches signals again

### ISSUE THREAD

**Issue by chengchengpei** [pika/pika#664]
When I run "sudo kill -USR1 <process_id>" or "sudo kill -USR2 <process_id>", Error 4 happened again.
The error info is "Exception while trying to read from AMQP [Errno 4] Interrupted system call". Similar problem reported here http://stackoverflow.com/questions/26282411/sighup-signal-give-exception-with-pika-in-python-2-7, but it is fixed in 0.9.14. Now, in 0.10, it appeares again.

**Comment by gmr** [comment]
Are you talking about the BlockingConnection adapter or pika as a whole?

**Comment by vitaly-krugl** [comment]
`libev_connection` appears to be the only adapter in pika that does anything with signals:
```
pika/adapters/libev_connection.py:import signal
pika/adapters/libev_connection.py:    If an on_signal_callback method is provided, the adapter creates signal
pika/adapters/libev_connection.py:    signal. See pyev/libev signal handling to understand why this is done.
pika/adapters/libev_connection.py:    :param on_signal_callback: Method to call if SIGINT or SIGTERM occur
pika/adapters/libev_connection.py:    :type on_signal_callback: method
pika/adapters/libev_connection.py:                 on_signal_callback=None):
pika/adapters/libev_connection.py:        :param on_signal_callback: Method to call if SIGINT or SIGTERM occur
pika/adapters/libev_connection.py:        :type on_signal_callback: method
pika/adapters/libev_connection.py:        self._on_signal_callback = on_signal_callback
pika/adapters/libev_connection.py:        LOGGER.debug('init io and signal watchers if any')
pika/adapters/libev_connection.py:        # reuse existing signal watchers, can only be declared for 1 ioloop
pika/adapters/libev_connection.py:            if self._on_signal_callback and not global_sigterm_watcher:
pika/adapters/libev_connection.py:                    self.ioloop.signal(signal.SIGTERM,
pika/adapters/libev_connection.py:            if self._on_signal_callback and not global_sigint_watcher:
pika/adapters/libev_connection.py:                global_sigint_watcher = self.ioloop.signal(signal.SIGINT,
pika/adapters/libev_connection.py:            if self._on_signal_callback:
pika/adapters/libev_connection.py:            if self._on_signal_callback:
pika/adapters/libev_connection.py:    def _handle_sigint(self, signal_watcher, libev_events):
pika/adapters/libev_connection.py:        """If an on_signal_callback has been defined, call it returning the
pika/adapters/libev_connection.py:        self._on_signal_callback('SIGINT')
pika/adapters/libev_connection.py:    def _handle_sigterm(self, signal_watcher, libev_events):
pika/adapters/libev_connection.py:        """If an on_signal_callback has been defined, call it returning the
pika/adapters/libev_connection.py:        self._on_signal_callback('SIGTERM')
```

**Comment by shinji-s** [comment]
@chengchengpei, if you mean IOError(Errno=4) by "Errno=4" you've likely hit the issue patched by this; 
https://github.com/pika/pika/commit/59dfa06cb18ce53e12979547312941eba04b0a6d

**Comment by vitaly-krugl** [comment]
@chengchengpei: which kernel are you using where this problem occurs? What's the `uname -a` ?

**Comment by chengchengpei** [comment]
Linux mypc 3.16.0-30-generic #40~14.04.1-Ubuntu SMP Thu Jan 15 17:43:14 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by chengchengpei** [comment]
@shinji-s Is this patch in 0.10.0?

**Comment by vitaly-krugl** [comment]
@chengchengpei, it's not in the release yet. 0.10.0 is at b907f91 (https://github.com/pika/pika/releases/tag/0.10.0), which predates @shinji-s's patch.

@chengchengpei, would you mind installing pika from master branch and verifying the fix? It would be great to get this patch released.

**Comment by chengchengpei** [comment]
@vitaly-krugl OK, I will test it today or tomorrow...

**Comment by chengchengpei** [comment]
I downloaded the source code, and run "python setup.py build; python setup.py install." start_consuming() can not be interupted by TERM, USR1, and USR2.

**Comment by vitaly-krugl** [comment]
@chengchengpei: thank you for testing it.

@gmr: Gavin, would you mind cutting a new release?

**Comment by chengchengpei** [comment]
@vitaly-krugl I have a question. I downloaded the source code, and run "python setup.py build; python setup.py install."  But I have no idea where pika is installed. and when I run my program, I am not sure which pika is used. I tried to find the new versrion of select_connnection.py, but failed in the path printed out when installing it.

**Comment by vitaly-krugl** [comment]
@chengchengpei: This works for me (in ipython):

```
In [1]: import pika

In [2]: pika.__version__
Out[2]: '0.10.0'

In [3]: pika
Out[3]: <module 'pika' from '/Users/vkruglikov/Library/Python/2.7/lib/python/site-packages/pika/__init__.pyc'>
```

**Comment by chengchengpei** [comment]
@vitaly-krugl my output is >>> import pika
>>> pika.__version__
'0.10.0'
>>> pika
<module 'pika' from '/home/myadmin/.virtualenvs/proj-dev/local/lib/python2.7/site-packages/pika-0.10.0-py2.7.egg/pika/__init__.pyc'>
>>>


/home/myadmin/.virtualenvs/porj-dev/local/lib/python2.7/site-packages/pika/adapters/select_connection.py looks like:

READ = 0x0001
WRITE = 0x0004
ERROR = 0x0008


if pika.compat.PY2:
    _SELECT_ERROR = select.error
else:
    # select.error was deprecated and replaced by OSError in python 3.3
    _SELECT_ERROR = OSError


It seems that the patch is not in my used pika....But when I installed pika, the output is:
Removing /home/myadmin/.virtualenvs/proj-dev/lib/python2.7/site-packages/pika-0.10.0-py2.7.egg
Copying pika-0.10.0-py2.7.egg to /home/myadmin/.virtualenvs/proj-dev/lib/python2.7/site-packages
pika 0.10.0 is already the active version in easy-install.pth

Installed /home/myadmin/.virtualenvs/proj-dev/lib/python2.7/site-packages/pika-0.10.0-py2.7.egg

**Comment by chengchengpei** [comment]
@vitaly-krugl Do you know what is the difference between /home/mysadmin/.virtualenvs/proj-dev/lib/python2.7/site-packages and /home/mysadmin/.virtualenvs/proj-dev/local/lib/python2.7/site-packages?

**Comment by chengchengpei** [comment]
@vitaly-krugl I am new to pika. I have two questions. (1) How to check whether one virtual host exists or not? (2) How to create it? I did not find hints. In most examples, vhost already exists..

**Comment by vitaly-krugl** [comment]
@chengchengpei, I am not familiar with the `.virtualenvs` directory hierarchy. I am guessing you or something on your system is using python virtualenv tools. 

In python, it's possible for a module to be installed in multiple places at the same time, including as root, `--user`, etc. When I want to make sure that the pika package is completely uninstalled, I run `pip uninstall pika` multiple times, until it tells me that there aren't any more. If pip was installed at the system level, you might get a permissions error message when uninstalling, in which case you need to re-run the command with `sudo pip uninstall pika`. After that, when you install pika, it should be the only version on your system.

**Comment by vitaly-krugl** [comment]
@chengchengpei, regarding your vhost checks and creation API questions, rabbitmq has a REST API that you can use, assuming you have the RabbitMQ Management Plugin installed (I don't remember if it gets installed by default, but if you can access your RabbitMQ broker's web GUI, then the management plugin should be in place). For example, if your rabbitmq broker is running on localhost, then you can get a summary of the API by browsing http://localhost:15672/api.

Here is an example of python code that creates a vhost: https://github.com/numenta/numenta-apps/blob/1ff572a21a5c27fd290822e572ce33f42e1ee19e/nta.utils/nta/utils/test_utils/amqp_test_utils.py#L145-L160

https://groups.google.com/forum/#!forum/rabbitmq-users is a good group to search and post.

**Comment by chengchengpei** [comment]
@vitaly-krugl pika/pika can not check or create vhost?

**Comment by vitaly-krugl** [comment]
> pika/pika can not check or create vhost?
@chengchengpei: I don't think so. I guess that the rationale is that pika implements the AMQP API, including RabbitMQ extensions. Vhosts and REST API are not part of AMQP.

**Comment by gmr** [comment]
Correct, you might want to check out http://pyrabbit.readthedocs.org for admin related behaviors

**Comment by vitaly-krugl** [comment]
Closing, fix is in master via PR #637, as noted and confirmed by @chengchengpei in earlier comments.

**Comment by hiqsol** [comment]
I also have run into this problem on Ubuntu 16.04 with pika 0.10.0
After installing pika from master branch everyting is ok!
So, release new version please :)
